semantic-analysis/src/Analysis/Syntax/Python.hs

```diff
@@ -1,82 +1,122 @@
+  where
+  suite = foldMap (subterms . Statement)
 
 
 -- Abstract interpretation
 
-eval0 :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m, Has (Reader Reference) sig m, Has S.Statement sig m) => Term -> m val
+eval0 :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m, Has (Reader Reference) sig m, Has S.Statement sig m, MonadFail m) => Term -> m val
 eval0 = fix eval
 
 eval
-  :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m, Has (Reader Reference) sig m, Has S.Statement sig m)
+  :: (Has (Env addr) sig m, HasLabelled Store (Store addr val) sig m, Has (Dom val) sig m, Has (Reader Reference) sig m, Has S.Statement sig m, MonadFail m)
   => (Term -> m val)
   -> (Term -> m val)
 eval eval = \case
-  T.Var n  -> lookupEnv n >>= maybe (dvar n) fetch
-  T.Term s -> case s of
-    Noop      -> dunit
-    Iff c t e -> do
-      c' <- eval c
-      dif c' (eval t) (eval e)
-    Bool b    -> dbool b
-    String s  -> dstring s
-    Throw e   -> eval e >>= ddie
-    Let n v b -> do
-      v' <- eval v
-      let' n v' (eval b)
-    t :>> u   -> do
-      t' <- eval t
-      u' <- eval u
-      t' >>> u'
-    Import ns -> S.simport ns >> dunit
-    Function n ps b -> letrec n (dabs ps (foldr (\ (p, a) m -> let' p a m) (eval b) . zip ps))
-    Call f as -> do
-      f' <- eval f
-      as' <- traverse eval as
-      dapp f' as'
-    Locate s t -> local (setSpan s) (eval t)
+  Module (Py.Module ss) -> suite ss
+  Statement (Py.Import is sp) -> setSpan sp $ do
+    for_ is $ \ Py.ImportItem{ Py.import_item_name = ns } -> case nonEmpty ns of
+      Nothing -> pure ()
+      Just ss -> S.simport (pack . Py.ident_string <$> ss)
+    dunit
+  Statement (Py.Pass sp) -> setSpan sp dunit
+  Statement (Py.Conditional cts e sp) -> setSpan sp $ foldr (\ (c, t) e -> do
+    c' <- eval (Expr c)
+    dif c' (suite t) e) (suite e) cts
+  Statement (Py.Raise (Py.RaiseV3 e) sp) -> setSpan sp $ case e of
+    Just (e, _) -> eval (Expr e) >>= ddie -- FIXME: from clause
+    Nothing     -> dunit >>= ddie
+    -- FIXME: RaiseV2
+    -- FIXME: whatever the second field is
+  Statement (Py.StmtExpr e sp) -> setSpan sp (eval (Expr e))
+  Statement (Py.Fun n ps _r ss sp) -> let ps' = mapMaybe (\case { Py.Param n _ _ _ -> Just (ident n) ; _ -> Nothing }) ps in setSpan sp $ letrec (ident n) (dabs ps' (foldr (\ (p, a) m -> let' p a m) (suite ss) . zip ps'))
+  Expr (Py.Var n sp) -> setSpan sp $ let n' = ident n in lookupEnv n' >>= maybe (dvar n') fetch
+  Expr (Py.Bool b sp) -> setSpan sp $ dbool b
+  Expr (Py.Strings ss sp) -> setSpan sp $ dstring (pack (mconcat ss))
+  Expr (Py.Call f as sp) -> setSpan sp $ do
+    f' <- eval (Expr f)
+    as' <- traverse eval (mapMaybe (\case { Py.ArgExpr e _ -> Just (Expr e) ; _ -> Nothing }) as)
+    -- FIXME: support keyword args &c.
+    dapp f' as'
+  _ -> fail "TBD"
```

Comment:
We handle one fewer case: `Let` is gone. However, `Let` was never produced from actual Python code using the TSG generator, so it's unclear why it was in there to begin with.

We also handle statement expressions which I think may not have been handled? And the failure case is explicit here, instead of being hidden in the term ingestion machinery like it had been.